keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt

```diff
@@ -60,34 +78,103 @@ class ClusterHandler(
+    imageId != null
+      && (this == null
+      || !this.isImageOnly()
+      || this.source.launchConfiguration.imageId != imageId)
+
+  private suspend fun Resource<Cluster>.maybeCreateServerGroupWithDynamicImage(resourceDiff: ResourceDiff<Cluster>?): Map<String, Any?> {
+    val imageId = this.imageIdFromMetadata()
+    return if (resourceDiff.shouldDeploy(imageId)) {
+      this.spec.createServerGroupJob() + mapOf("amiName" to imageId!!)
+    } else {
+      emptyMap()
+    }
+  }
+
+  override fun upsert(resource: Resource<Cluster>, resourceDiff: ResourceDiff<Cluster>?) {
+    runBlocking {
       val spec = resource.spec
       val job = when {
-        diff.isCapacityOnly() -> spec.resizeServerGroupJob()
-        else -> spec.createServerGroupJob()
+        resourceDiff.isCapacityOnly() -> spec.resizeServerGroupJob()
+        spec.hasStaticImage -> spec.createServerGroupJob()
+        !resourceDiff.isImageOnly() -> emptyMap()
+        else -> resource.maybeCreateServerGroupWithDynamicImage(resourceDiff)
       }
 
-      log.info("Upserting cluster using task: {}", job)
+      if (job.isNotEmpty()) {
+        log.info("Upserting cluster using task: {}", job)
 
-      orcaService
-        .orchestrate(OrchestrationRequest(
-          "Upsert cluster ${spec.moniker.cluster} in ${spec.location.accountName}/${spec.location.region}",
-          spec.moniker.application,
-          "Upsert cluster ${spec.moniker.cluster} in ${spec.location.accountName}/${spec.location.region}",
-          listOf(Job(job["type"].toString(), job)),
-          OrchestrationTrigger(resource.metadata.name.toString())
-        ))
-        .await()
+        orcaService
+          .orchestrate(OrchestrationRequest(
+            "Upsert cluster ${spec.moniker.cluster} in ${spec.location.accountName}/${spec.location.region}",
+            spec.moniker.application,
+            "Upsert cluster ${spec.moniker.cluster} in ${spec.location.accountName}/${spec.location.region}",
+            listOf(Job(job["type"].toString(), job)),
+            OrchestrationTrigger(resource.metadata.name.toString())
+          ))
+          .await()
+      } else {
+        null
+      }
+    }?.also {
+      log.info("Started task {} to upsert cluster", it.ref)
     }
-    log.info("Started task {} to upsert cluster", taskRef.ref)
   }
 
   /**
    * @return `true` if the only changes in the diff are to capacity.
    */
-  private fun DiffNode?.isCapacityOnly(): Boolean =
-    this != null && affectedRootPropertyTypes.all { it == Capacity::class.java }
+  private fun ResourceDiff<Cluster>?.isCapacityOnly(): Boolean =
+    this != null && diff.affectedRootPropertyTypes.all { it == Capacity::class.java }
+
+  private fun ResourceDiff<Cluster>?.isImageOnly(): Boolean =
+    this != null && this.diff.imageChanged && this.diff.nodePaths.size == 1
+
+  private val imageIdPath = NodePath.with("launchConfiguration", "imageId")
+
+  private val DiffNode.imageChanged: Boolean
```

Comment:
feels like we might want to think about how we could hide away this object diffing library, I suspect we'll end up with many similar bits of code to this as we keep building out keel, it might be nice if we had some layer in front of it to contain the dependency

Reply:
I agree